### PR TITLE
Initial GDPS UU layers implementation

### PR DESCRIPTION
**Pull request by Dukestep**
This PR contains an implementation for managing GeoMet GDPS layers that require having two or more variables in the registry before creating the tileindex record for the given layer.

Steps taken to enable this functionality:

1. Modified the `model_gem_global.yml` configuration file to add a `dependencies` key that contains the name of the GeoMet layer on which the variable depends on. Also added a `bands_order` key that specifies the order in which the bands need to be created within the VRT. For example:
```yaml
        UGRD_ISBL_1000:
            [...]
            geomet_layers:
                GDPS.PRES_UGRD.1000:
                    forecast_hours: 000/240/PT3H
                GDPS.PRES_UU.1000:
                    dependencies:
                        - GDPS.PRES_VGRD.1000
                    forecast_hours: 000/240/PT3H
            bands_order:
                - UGRD_ISBL
                - VGRD_ISBL
```
2. When `model_gem_global.py` encounters a layer with dependencies, it checks that all dependencies listed are available in the tileindex for the same model run and forecast hour (see [L127](https://github.com/ECCC-MSC/geomet-data-registry/compare/master...Dukestep:uu-layers-gdps?expand=1#diff-5770a3fccda74a8fe7a62c1ef12135cbR127)). If all dependencies are found, items are created for this layer. The `wx_variable` property of the item is also modified to include the names of all variables used by the GeoMet layer. A VRT is created using the filepaths of the current variable and those of its dependencies. The `bands_order` for the variable can be passed to `VRTDataset` object in order to create the VRT with the bands in the specified order.

3. Time keys for the layer are only updated when all the layer's dependencies all have the same `default_model_run` value (see `model_gem_global.py` [L201](https://github.com/ECCC-MSC/geomet-data-registry/compare/master...Dukestep:uu-layers-gdps?expand=1#diff-5770a3fccda74a8fe7a62c1ef12135cbR201)).

Open to any suggestions and improvements. Keeping this as a WIP until we agree with the implementation.

This has been tested with a full GDPS model run and I have not identified any glaring issues.

### CODE REVIEW

File: geomet_data_registry/layer/base.py [hunk 1]
```diff
@@ -189,6 +197,64 @@ def update_count(self, item, r, item_dict):
...
+        Checks if all layer dependencies are available in the tileindex
+        for a given model run and forecast hour.
+        :param layers_list: `list` of layer dependencies
+        :param str_mr: `str` of model run
+        :param str_fh: `str` of forecast hour
+        :returns: `list` of GeoJSON objects for retrieved dependencies
+        """
+        dependencies = [self.tileindex.get('{}-{}-{}'.format(layer, str_mr, str_fh)) for layer in layers_list]  # noqa
+
+        if None not in dependencies:
```
Comment (tomkralidis) [inline geomet_data_registry/layer/base.py]: What happens if `None` is in `dependencies`?  Should we return an empty list ?

File: geomet_data_registry/layer/base.py [hunk 2]
```diff
@@ -189,6 +197,64 @@ def update_count(self, item, r, item_dict):
...
+                                       ['filepath'] for dependency
+                                       in dependencies]
+        vrt = VRTDataset(
+            filepaths,
+            raster_x_size=image_dimension['x'],
+            raster_y_size=image_dimension['y'],
+            bands_order=bands_order).build()
+
+        weather_variables = '{},{}'.format(
+            self.wx_variable, ",".join(
```
Comment (tomkralidis) [inline geomet_data_registry/layer/base.py]: single quotes

File: geomet_data_registry/layer/base.py [hunk 3]
```diff
@@ -189,6 +197,64 @@ def update_count(self, item, r, item_dict):
...
+
+        return vrt, weather_variables
+
+    def check_dependencies_default_mr(self, mr_datetime, dependencies):
+        """
+        For each dependency, verify that a default model run is available in
+        the store and the value to the passed model run time.
+        :param dependencies: `list` of dependencies to check
+        :param mr_datetime: `datetime` to compare dependencies to
+        :returns: True if each dependency's default model run is equal to
```
Comment (tomkralidis) [inline geomet_data_registry/layer/base.py]: Should this be `bool` of whether each dependency's default model run is equal to the `mr_datetime` parameter ?

File: geomet_data_registry/layer/model_gem_global.py [hunk 4]
```diff
@@ -174,6 +194,16 @@ def add_time_key(self):
                              "Not updating time keys.".format(default_model_run, stored_default_model_run))  # noqa
                 continue
 
+            if 'dependencies' in values:
+                if not self.check_dependencies_default_mr(
+                        self.date_, values['dependencies']):
+                    LOGGER.debug(
+                        "The default model run for at least one "
```
Comment (tomkralidis) [inline geomet_data_registry/layer/model_gem_global.py]: Single quotes.

File: geomet_data_registry/util.py [hunk 5]
```diff
@@ -26,6 +26,47 @@
 DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
 
 
+class VRTDataset:
```
Comment (tomkralidis) [inline geomet_data_registry/util.py]: `class VRTDataset(object):`

File: geomet_data_registry/util.py [hunk 6]
```diff
@@ -26,6 +26,47 @@
...
+        self.vrt_raster_band_template = '''
+<VRTRasterBand dataType="Byte" band="{}">
+    <ComplexSource>
+        <SourceFilename>{}</SourceFilename>
+        <SourceBand>1</SourceBand>
+        <ScaleOffset>0.0</ScaleOffset>
+        <ScaleRatio>1.0</ScaleRatio>
+    </ComplexSource>
+</VRTRasterBand>
+        '''.replace('\n', '')
```
Comment (tomkralidis) [inline geomet_data_registry/util.py]: are you collapsing the string here into one line?  Perhaps add a specific class function instead?

File: geomet_data_registry/util.py [hunk 7]
```diff
@@ -26,6 +26,47 @@
...
+        self.filepaths = filepaths
+        self.bands_order = bands_order
+
+    def build(self):
```
Comment (tomkralidis) [inline geomet_data_registry/util.py]: docstring?

File: geomet_data_registry/util.py [hunk 8]
```diff
@@ -26,6 +26,47 @@
...
+        if self.bands_order:
+            self.filepaths = sorted(self.filepaths, key=lambda fp:
+                                    self.sort_bands(fp))
+        for index, filepath in enumerate(self.filepaths, start=1):
+            self.vrt_dataset.insert(-1, self.vrt_raster_band_template.format(
+                index, filepath))
+        return ''.join(self.vrt_dataset)
+
+    def sort_bands(self, filepath):
```
Comment (tomkralidis) [inline geomet_data_registry/util.py]: docstring?

File: geomet_data_registry/util.py [hunk 9]
```diff
@@ -26,6 +26,47 @@
...
+        return [idx for idx, value in enumerate(self.bands_order) if value
```
Comment (tomkralidis) [inline geomet_data_registry/util.py]: - Log what's going on here?
- how does the else condition evaluate?

File: geomet_data_registry/util.py (hunk 6 above)
Comment (Dukestep) [reply geomet_data_registry/util.py]: Added a `VRTDatateset.formatted_vrt()` method to format the VRT appropriately (single line with no whitespace formatting).

File: geomet_data_registry/util.py (hunk 7 above)
Comment (Dukestep) [reply geomet_data_registry/util.py]: Done!

File: geomet_data_registry/util.py (hunk 8 above)
Comment (Dukestep) [reply geomet_data_registry/util.py]: Done, sorry!

File: geomet_data_registry/layer/base.py (hunk 3 above)
Comment (Dukestep) [reply geomet_data_registry/layer/base.py]: Yes, modified the docstring to make this more clear.

File: geomet_data_registry/util.py (hunk 9 above)
Comment (Dukestep) [reply geomet_data_registry/util.py]: The sort bands method now returns a `VRTDatasetError` if none of the `self.bands_order` values can be found in the provided filepath, meaning that the filepaths cannot be properly sorted prior to the VRT construction.

I also added a log debug message indicating that the filepath is being checked against the provided bands order.

**Pull request merged**